### ISSUE TITLE
"ValueError: Buffer has wrong number of dimensions (expected 4, got 2)" when using dw_bc_newton term

### ISSUE THREAD

**Issue by fanoway** [sfepy/sfepy#817]
I am trying to mode l a combination heat generation, diffusion problem with a convective bnoundary layer. I am trying to use the dw_bc_nerwton term to capture the convection on the edges. I am encountering the following error however when trying to solve.

`Traceback (most recent call last):
  File "C:\Users\User Name\Downloads\sfep_model\app1.py", line 268, in <module>
    model(show_regions=False, show_results=True)
  File "C:\Users\User Name\Downloads\sfep_model\app1.py", line 230, in model
    variables = pb.solve(
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\discrete\problem.py", line 1429, in solve
    vec = tss(variables.get_state(self.active_only, force=True),
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\solvers\ts_solvers.py", line 32, in _standard_ts_call
    result = call(self, vec0=vec0, nls=nls, init_fun=init_fun,
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\solvers\ts_solvers.py", line 73, in __call__
    vec = nls(vec0)
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\solvers\nls.py", line 260, in __call__
    vec_r = fun(vec_x)
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\discrete\evaluate.py", line 73, in eval_residual
    vec_r = self.problem.equations.eval_residuals(vec)
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\discrete\equations.py", line 674, in eval_residuals
    self.evaluate(mode='weak', dw_mode='vector', asm_obj=out)
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\discrete\equations.py", line 606, in evaluate
    out = eq.evaluate(mode=mode, dw_mode=dw_mode,
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\discrete\equations.py", line 829, in evaluate
    val, iels, status = term.evaluate(mode=mode,
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\terms\terms.py", line 1481, in evaluate
    vals, status = self.eval_real(shape, fargs, mode,
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\terms\terms.py", line 1347, in eval_real
    status = self.call_function(out, fargs)
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\terms\terms.py", line 1324, in call_function
    status = self.function(out, *fargs)
  File "C:\Users\User Name\miniconda3\envs\heat_model\lib\site-packages\sfepy\terms\terms_dot.py", line 72, in dw_dot
    status = fun(out, mat, val_qp, vgeo, sgeo, fmode)
  File "sfepy\terms\extmods\terms.pyx", line 1545, in sfepy.terms.extmods.terms.dw_volume_dot_scalar
  File "sfepy\discrete\common\extmods\_fmfield.pyx", line 5, in sfepy.discrete.common.extmods._fmfield.array2fmfield4
ValueError: Buffer has wrong number of dimensions (expected 4, got 2)
(heat_model) PS C:\Users\User Name\Downloads\sfep_model>
`

Code I am trying to run

`import meshio
import numpy as np
import pygmsh
import pyqtgraph as pg
from sfepy.base.base import IndexedStruct
from sfepy.discrete import (
    Equation,
    Equations,
    FieldVariable,
    Function,
    Integral,
    Material,
    Problem,
)
from sfepy.discrete.conditions import Conditions, EssentialBC
from sfepy.discrete.fem import FEDomain, Field, Mesh
from sfepy.discrete.probes import LineProbe
from sfepy.postprocess.viewer import Viewer
from sfepy.solvers.ls import ScipyDirect
from sfepy.solvers.nls import Newton
from sfepy.terms import Term

def create_mesh():
    with pygmsh.geo.Geometry() as geom:
        geom.add_polygon(
            [
                [0.0, 0.0],
                [0, 1],
                [1, 1],
                [1.2, 0.7],
                [1.2, 0],
            ],
            mesh_size=0.2,
        )
        mesh = geom.generate_mesh()

    return mesh

def gen_probe(problem, x_range, y_range):
    """
    Define a LineProbe at halfway in y across all of x
    """

    # Use enough points for higher order approximations.
    n_point = 1000

    p0, p1 = np.array([x_range[0], 0.5 * sum(y_range)]), np.array(
        [x_range[1], 0.5 * sum(y_range)]
    )
    probe = LineProbe(p0, p1, n_point, share_geometry=True)
    # Workaround current probe code shortcoming.
    probe.set_options(close_limit=0.5)

    label = ["%s -> %s" % (p0, p1)]

    return probe, label

def probe_results(win, t, probe, label):
    """
    Probe the results using the given probe and plot the probed values.
    """

    results = {}

    pars, vals = probe(t)
    results["t"] = (pars, vals[:, 0])
    pars, vals = results["t"]

    pgplt = win.addPlot()
    pgplt.clear()

    pgplt.addLegend()
    pgplt.plot(pars, vals, name=f"T, Probe = {label}", pen="r")
    pgplt.setLabel("left", "Temperature")
    pgplt.setLabel("bottom", "x position")

    return win, results

def model(show_regions=False, show_results=False):
    mesh = Mesh.from_file("working_mesh.vtk")

    domain = FEDomain("domain", mesh)
    x_range = domain.get_mesh_bounding_box()[:, 0]
    y_range = domain.get_mesh_bounding_box()[:, 1]

    eps = 1e-8 * (x_range[1] - x_range[0])

    omega = domain.create_region("Omega", "all")

    Outer = domain.create_region(
        "Gamma1", "vertices in x < %.10f" % (x_range[0] + eps), "facet"
    )
    Inner = domain.create_region(
        "Gamma2", "vertices in x > %.10f" % (x_range[1] - eps), "facet"
    )

    field = Field.from_args("temperature", np.float64, 1, omega, approx_order=2)

    t = FieldVariable("t", "unknown", field)
    s = FieldVariable("s", "test", field, primary_var_name="t")

    def cu_func(ts, coors, problem, mode=None, **kwargs):
        """
        Returns the thermal diffusivity, and the power of the copper.
        """
        # TODO set power based on temperature

        if mode == "qp":

            power_per_volume = 25

            power = np.empty((coors.shape[0], 1, 1), dtype=np.float64)
            power.fill(power_per_volume)

            diffusivity = np.array([111 * np.eye(2)])
            diffuse_array = diffusivity

            for _ in range(coors.shape[0] - 1):
                diffuse_array = np.vstack((diffuse_array, diffusivity))

            return {"power": power, "diffusivity": diffuse_array}

    cu_fun = Function(
        "cu_func",
        cu_func,
    )

    cu = Material("cu", function=cu_fun)

    air = Material(
        "air",
        values={".h": np.array([[50e-6]]), ".T": np.array([[50]])},
    )
    joint = Material(
        "joint",
        values={".h": np.array([[50e-6]]), ".T": np.array([[50]])},
    )

    integral = Integral("i", order=2)

    diffusion = Term.new(
        "dw_diffusion(cu.diffusivity, s, t)", integral, omega, cu=cu, s=s, t=t
    )
    heat_gen = Term.new("dw_integrate(cu.power, s)", integral, omega, cu=cu, s=s)

    convection = Term.new(
        "dw_bc_newton(joint.h,air.T, s,t)",
        integral,
        omega,
        joint=joint,
        air=air,
        s=s,
        t=t,
    )

    eq = Equation("thermal", diffusion + convection - heat_gen)
    eqs = Equations([eq])

    # TODO change EBC to be heat convection from surface

    ls = ScipyDirect({})
    nls_status = IndexedStruct()
    nls = Newton({}, lin_solver=ls, status=nls_status)

    pb = Problem("heat", equations=eqs)

    if show_regions:
        pb.save_regions_as_groups("regions")
        view = Viewer("regions.vtk")
        view(is_wireframe=True)

    pb.set_solver(nls)

    status = IndexedStruct()
    variables = pb.solve(
        status=status,
    )

    print("\n######")
    print("Nonlinear solver status:\n", nls_status)
    print("\n######")
    print("Stationary solver status:\n", status)
    print("######")

    out = variables.create_output()
    if show_results:
        pb.save_state("thermal.vtk", out=out)
        view = Viewer("thermal.vtk")
        view(is_wireframe=True, is_scalar_bar=True)

    probe, label = gen_probe(pb, x_range, y_range)

if __name__ == "__main__":

    mesh = create_mesh()

    meshio.write(
        "working_mesh.vtk",
        meshio.Mesh(points=mesh.points, cells={"triangle": mesh.cells[1].data}),
    )

    model(show_regions=False, show_results=True)
`

**Comment by rc** [comment]
Correct the following (notice no dots in the dict keys):
```
    air = Material(
        "air",
        values={"h": np.array([[50e-6]]), "T": np.array([[50.]])},
    )
    joint = Material(
        "joint",
        values={"h": np.array([[50e-6]]), "T": np.array([[50.]])},
    )
```
or simply:
```
    air = Material(
        "air", h=50e-6, T=50,
    )
    joint = Material(
        "joint", h=50e-6, T=50,
    )
```
and then use a surface region in the dw_bc_newton term:
```
    convection = Term.new(
        "dw_bc_newton(joint.h,air.T, s,t)",
        integral,
        Outer,
        joint=joint,
        air=air,
        s=s,
        t=t,
    )
```
With the above fixes your code runs for me.

**Comment by fanoway** [comment]
Thanks, I made the changes you recommended and I get this error instead.

`ValueError: operands could not be broadcast together with shapes (19,2,1,1) (1,3,1,1)`

I noticed I didn't have the facet flag set in my get region lines. Adding that throws this error instead

`  File "C:\Users\user.name\Miniconda3\envs\gen_modeller\lib\site-packages\sfepy\terms\terms_dot.py", line 194, in get_fargs
    fargs = fargs[:1] + (fargs[1] - p_outer) + fargs[2:]
TypeError: unsupported operand type(s) for +: 'float' and 'sfepy.discrete.common.extmods.mappings.CMapping'`

Note, I'm using windows 7, I have tested on windows 10 and I do not have this issue.
